### ISSUE TITLE
Youtube Previews not working after upgrade

### ISSUE THREAD

**Issue by swaggboi** [mattermost/mattermost#16429]
#### Summary
Youtube Previews not working after upgrade

#### Steps to reproduce
Paste URL to some YouTube video in a channel

#### Expected behavior
YouTube previews should work

#### Observed behavior (that appears unintentional) 
I have recently upgraded from 5.26 to 5.29 (using mattermost/mattermost-docker) and now YouTube previews do not work on either a desktop client (tested macOS, Linux & Windows) or a desktop web browser (tested Safari and Firefox). Curiously they DO work in the mobile application (tested iOS and Android clients). On the desktop, if I scroll back a few days before the upgrade occurred, the previews are still in place for those old posts.

#### Possible fixes
I will list things I have tried after reading previous posts that looked like they may be similar:
- System Console > SITE CONFIGURATION > Posts; ensured "Enable Link Previews" is set to True and ensured Google API key is populated
- Cut new Google API key
- Accessed shell of the app container; performed an `nslookup www.youtube.com` and `curl -k -I -L https://www.youtube.com` which returned IP addresses and a 200 response, respectively
- Ensured photo proxy is off as many suggestions referenced this; it was never on and should not be as I am not using an outbound proxy
- Checked `diff` of config.json; nothing sticks out to me https://pastebin.com/9EHtuqFe
- Enabled DEBUG level logging; tail'd the log file while posting YouTube links. Didn't see anything relevant printed

**Comment by amyblais** [comment]
@swaggboi This Github issue sounds similar and they posted a resolution here: https://github.com/mattermost/mattermost-server/issues/15604#issuecomment-703893820. Could any of the points there help?

**Comment by swaggboi** [comment]
@amyblais yes, I saw this one already which is why I ran the `nslookup` and `curl`. For me name resolution works fine and `curl` returns a 200 rather than a 302

**Comment by gittigger** [comment]
I reported #15604 - I don't think the problem is the upgrade, or the issue I saw on this ticket. These previews now do not work on my setup at all in the same way @swaggboi has reported. I have changed nothing on my setup since that ticket.  
@amyblais Has anyone else reported this anywhere?

**Comment by hmhealey** [comment]
The difference between the web/desktop and mobile apps is likely because for the web/desktop apps, the server is responsible for generating those previews while the mobile app generates them itself, but issues with the server requesting previews are usually diagnosed by checking with curl like you already did.

Do any other non-YouTube links generate previews like by posting https://mattermost.com or a direct link to an image?
![Screen Shot 2020-12-03 at 1 48 33 PM](https://user-images.githubusercontent.com/3277310/101074225-84770780-356e-11eb-8796-7261ab0622fd.png)

**Comment by swaggboi** [comment]
Yes so far all other links generate previews as expected. Tested with the URL you provided:
![Screen Shot 2020-12-03 at 13 54 10](https://user-images.githubusercontent.com/38310112/101074790-2991e000-356f-11eb-9542-b866b0c37ff6.jpg)

**Comment by amyblais** [comment]
@hmhealey ^

**Comment by hmhealey** [comment]
Strange. YouTube previews work on our community server which runs a few different versions, but I can't get them to work anywhere else. Hopefully we're not getting blocked by YouTube for some reason.

I've filed a Jira ticket to investigate further since this is probably something we'll want fixed quickly. Here's a link to that ticket in case you'd like to track it in more detail: 
https://mattermost.atlassian.net/browse/MM-31171

**Comment by hmhealey** [comment]
I've posted an update in the Jira ticket, but just to summarize it here, YouTube has started sometimes putting the OpenGraph metadata that we use to generate these previews into the wrong place in their HTML which causes the library we're using to scrape it to not find it. It seems inconsistent, but we're seeing this happen more frequently as time goes on which breaks YouTube previews on more and more servers.

**Comment by amyblais** [comment]
Closing as this is queued to be fixed in upcoming v5.30 release,

**Comment by hmhealey** [comment]
@goeffthomas Thanks for the report! We're currently tracking the issue with MM-60351 if you want to know more. Unfortunately, this is an issue on YouTube's end, so it'll likely affect every version of MM.

Are you able to provide any information about where your servers are being hosted? I'm curious if they're self-hosted or if they're hosted with a cloud provider.

For some reason, YouTube seems to have stopped giving us the data that we need entirely, but it only affects certain servers at the moment. Currently, our cloud servers have stopped showing YouTube previews, but they're working correctly for my development server, so we have some theories about YouTube not sending metadata to servers that are hosted on AWS

**Comment by goeffthomas** [comment]
> Are you able to provide any information about where your servers are being hosted? I'm curious if they're self-hosted or if they're hosted with a cloud provider.

We host ours on GCP, so I think that rules out the AWS theory.

**Comment by daydr3am3r** [comment]
Just wanted to chime in and report I have the same issue since August 15th, and my instance is hosted in Linode Frankfurt.

**Comment by hmhealey** [comment]
Thanks all for the details, everyone. There's still a chance they're targeting these changes at cloud providers in general since the previews still work just fine on peoples' machines from what we can tell, but I think we're likely going to have to find a different way to do these previews going forward. We've got a few different options like using YouTube's API directly instead of relying on OpenGraph metadata, and we're discussing which of those we're going to go with

**Comment by daydr3am3r** [comment]
@hmhealey thank you for the update. In the meantime, maybe it's useful to test the hypothesis in whichever way we can, namely:

1. Can you share the user agent string Mattermost uses? We can use curl or similar tools with that user agent string to see if the request is indeed blocked
2. Can a proxy be used for the requests towards Youtube? e.g. would an image proxy work?

**Comment by hmhealey** [comment]
I've got good news! A coworker pointed out to me that we can use oEmbed to get the information required for the YouTube preview instead of relying on OpenGraph, and that ended up being much easier to implement than anything I'd thought of myself 😅.

You can take a look at #28312 if you're curious for more information, but this'll at the very least be fixed in 10.2 in November, but I'm going to see about getting it backported further if possible

**Comment by hmhealey** [comment]
@daydr3am3r In case you're still at all curious:

1. Mattermost uses `Mattermost-Bot/1.0` when making requests to other servers for things like link metadata. I found that using either that or curl's default user agent would cause YouTube to not return the required meta tags when I was on any affected server. If I used Facebook's, it worked
2. You could potentially do it with something like mitmproxy that'd also let you override the user agent, but whatever's doing that would likely be pretty janky. We could theoretically add support for something equivalent to the image proxy, but that would need changes to MM to support

**Comment by daydr3am3r** [comment]
@hmhealey thank you for the information. 
I used the following curl config file from my Linode Frankfurt VPS to emulate the user agent, and I think it seemed to work, any idea as to why? I am also attaching the output.

```
url = "https://www.youtube.com/watch?v=5ZGIQDcLkCI"
output = "youtube-curl.html"
user-agent = "Mattermost-Bot/1.0"
```

**Comment by hmhealey** [comment]
You would actually need to look at the HTML in the response to know if it'll generate link previews. We use [OpenGraph](https://ogp.me/) for most of our link previews which stores its information in meta tags like `<meta property="og:title" content="Some video" />`. If you look at the `youtube-curl.txt` that you posted and search for `og:`, those tags are all missing, but if you run the same command from a personal machine, you'll find those there

**Comment by leetosc** [comment]
> I've got good news! A coworker pointed out to me that we can use oEmbed to get the information required for the YouTube preview instead of relying on OpenGraph, and that ended up being much easier to implement than anything I'd thought of myself 😅.
> 
> You can take a look at #28312 if you're curious for more information, but this'll at the very least be fixed in 10.2 in November, but I'm going to see about getting it backported further if possible

Please keep us updated if the fix can be applied to v9 as well. We are not ready to upgrade to v10 yet. Thanks!

**Comment by daydr3am3r** [comment]
@leetosc : PR has been merged for v9.5: https://github.com/mattermost/mattermost/pull/28375

**Comment by hmhealey** [comment]
Yep, that's right. Along with 10.1.0, this'll also be fixed in the 9.11.3 and 9.5.11 ESR releases which should all be coming out in by next week
